### PR TITLE
Augmented blueprint EP031 `telemetry_int`

### LINKED ISSUE #400: placeholder to augment blueprint EP031 (`telemetry_int`)
This is a placeholder to augment blueprint EP031 (`telemetry_int`):

- Update certain endpoints that have evolved based on https://github.com/kytos-ng/telemetry_int/blob/master/main.py
- Further specify how consistency check will be covered
- Confirm minor implementation details (that still haven't been mapped as issues) and expected behavior

### PR DESCRIPTION
**Pull request by viniarck**
Closes #400 

### Summary

As I was breaking down the `epic_telemetry_int_v1` epic, I realized that certain minor requirements still needed further refinement or confirmation, so I went ahead and augmented here accordingly. Also, I updated the rest endpoints accordingly based on the initial prototype that @jab1982 has developed. cc'ing @jab1982 and @italovalcy to confirm and discuss if the requirements are as expected for this v1 iteration.

### CODE REVIEW

File: docs/blueprints/EP031.rst
Comment (viniarck) [inline docs/blueprints/EP031.rst:252]: `undeployed` is needed if it ever gets disabled via an update.

File: docs/blueprints/EP031.rst
Comment (viniarck) [inline docs/blueprints/EP031.rst:253]: This is for keeping track of `proxy_interface` status changes, and also `proxy_port` metadata changes

File: docs/blueprints/EP031.rst
Comment (viniarck) [inline docs/blueprints/EP031.rst:282]: Since we have certain events or asynchronous side effects that can happen in addition to a HTTP request from clients, then we can use `"status_reason"` here too to fill out any error (they'll be specified as we go, but basically they'll be enum string values).

File: docs/blueprints/EP031.rst
Comment (viniarck) [inline docs/blueprints/EP031.rst:299]: We'll need to discuss if for production you'll need the consistency check completely implemented too, on `sdntrace_cp` these would also be needed:

https://github.com/kytos-ng/sdntrace_cp/issues/103

https://github.com/kytos-ng/sdntrace_cp/issues/104

File: docs/blueprints/EP031.rst
Comment (viniarck) [inline docs/blueprints/EP031.rst:317]: Looks like there was a possibility of not using loops for inter EVCs as mentioned [on EP033](https://github.com/kytos-ng/kytos/blob/master/docs/blueprints/EP033.rst), but in the initial prototype both (inter and intra) were requiring a loop, so I'm assuming that'd be case, let me know if I misunderstood or if indeed INT will always require a loop for both cases

File: docs/blueprints/EP031.rst
Comment (italovalcy) [reply docs/blueprints/EP031.rst:299]: Yes, we will need to have consistency check completely implemented because otherwise, we can misclassify an EVC with INT enabled as working when it might not be fully functional.

Comment (italovalcy) [reply docs/blueprints/EP031.rst:299]: Additionally, the consistency should also check if the proxy port is active and take actions accordingly.

File: docs/blueprints/EP031.rst
Comment (italovalcy) [inline docs/blueprints/EP031.rst:313]:
```suggestion
- Once ``mef_eline`` creates an EVC, it can optionally request that INT be provisioned. For this case, a ``telemetry_request: dict`` needs to be set in the metadata, currently no keys are needed, but as more options are supported in the future, they can be set. If ``telemetry_int`` can't provision ``telemetry_int``, then it'll set the ``telemetry: {"status": "disabled", "status_reason": ["<reason>"]}`` metadata, updating the status and filling out the reason accordingly.
```

**Review by italovalcy — APPROVED** [review]
Excellent points, Vinicius! thanks for sending the updates. I've made a few comments, please let me know if they make any sense for you.

File: docs/blueprints/EP031.rst
Comment (viniarck) [reply docs/blueprints/EP031.rst:251]: It'll be covered by `kytos.topology.updated`, since ultimately an interruption will also result in an `kytos.topology.updated` being published.

Comment (viniarck) [reply docs/blueprints/EP031.rst:251]: Right, indeed, initially I was assuming that it'd behave like a down UNI, since I hoping that eventually it'd probably be going up again since it would still exist: 

> If a proxy port source or destination status changes to DOWN or DISABLED (or if the loop stoops being a loop) all of associated EVCs should be considered not active, similar to what has been implemented when a UNI changes its UP state. No flows should be removed. Currently, mef_eline doesn't allow an EVC to be deactivated, so this will need to be published via events, depending if an EVC also starts to use status and status_reason then interruptions can also be used.

But, indeed, it could actually just remove the flows and allow the traffic to fallback to `mef_eline`. It's to you @italovalcy and @jab1982 to decide the behavior you'd like to see here, flow removals and then eventually trying to deploy it again once it goes up or consistency results in positive might be the the smoothest option, what do you think?

File: docs/blueprints/EP031.rst
Comment (viniarck) [reply docs/blueprints/EP031.rst:264]: Yes

File: docs/blueprints/EP031.rst
Comment (viniarck) [reply docs/blueprints/EP031.rst:266]: Right. Yes, thinking about it again, probably a `PATCH /telemetry_int/v1/evc/redeploy body evc_ids: []` would be handy, otherwise as you said you'd have to `/disable force` and then `/enable force`. I'll update the doc shortly, let me know if you'd be OK with it, it'd behave the same a `mef_eline` `redeploy`

File: docs/blueprints/EP031.rst
Comment (viniarck) [reply docs/blueprints/EP031.rst:299]: > Yes, we will need to have consistency check completely implemented because otherwise, we can misclassify an EVC with INT enabled as working when it might not be fully functional.

Right. So, we'll need to include this extra effort on `sdntrace_cp` then. @gretelliz let me know if you could collaborate on that part and if you think it's feasible, this epic here has high priority for the `2023.2` version it doesn't need immediate action, but should be on the top of the list.

Comment (viniarck) [reply docs/blueprints/EP031.rst:299]: > Additionally, the consistency should also check if the proxy port is active and take actions accordingly.

@italovalcy @jab1982, we'll need to specify here the expected behavior if the consistency check never succeeds, `mef_eline` worst case it to keep trying to redeploy, but `telemetry_int` in the worst case, should it remove the flows and give up to fallback to `mef_eline`? and then only try to activate/redeploy again if the EVC is active but also proxy ports are all set?

File: docs/blueprints/EP031.rst
Comment (viniarck) [reply docs/blueprints/EP031.rst:313]: Thanks, Italo, let me apply this patch.

File: docs/blueprints/EP031.rst
Comment (viniarck) [reply docs/blueprints/EP031.rst:241]: Sure thing, yes, let's double check it:

- [`DELETE /DELETE /v2/evc/{circuit_id}` publishes `deleted`](https://github.com/kytos-ng/mef_eline/blob/master/main.py#L413) 
- [When it fails to redeploy due to link down publishes `error_redeploy_link_down`](https://github.com/kytos-ng/mef_eline/blob/master/main.py#L802)

Good catch, let's include them just so flows can get deleted too.

Comment (viniarck) [reply docs/blueprints/EP031.rst]: Updated here 01d81216de904f8f8ce92ec59b74517923ac5a13

File: docs/blueprints/EP031.rst
Comment (viniarck) [reply docs/blueprints/EP031.rst:266]: Included here 2fdcbdb7909c480383218ef8a4cb2f81f31aaec1

File: docs/blueprints/EP031.rst
Comment (viniarck) [reply docs/blueprints/EP031.rst:251]: If we were to support this, we'd also need to keep track of the looped links and which EVCs are going over it and react accordingly, we'll need to keep track of proxy ports anyway so that wouldn't be too far away. Good thing that we do have the link timer up to also avoid too much unnecessary IO with FlowMods. When it goes up it could redeploy again. I'll wait for your confirmation, but indeed looks like falling back would be the optimal (and less surprising) approach.

File: docs/blueprints/EP031.rst
Comment (viniarck) [inline docs/blueprints/EP031.rst:42]: cc'ing @italovalcy @jab1982 regarding the requirement to support also EPL on table `"n"`. I left the requirement still higher level, but looks like we'll probably go for another table just for EPLs. I'll leave it up to you guys to confirm, since this is highly dependent on pipeline configuration architecture. The implementation to support another table shouldn't be too difficult, although if we still had a single one it'd be even easier to maintain and test.

If we do end up introducing a new table, we should also think about the concept of table group, that we introduced on [`of_multi_table` NApp](https://github.com/kytos-ng/of_multi_table/blob/master/openapi.yml#L164-L169), which in summary is reponsible for categorizing a set of flows that should always be installed in a particular table. So far `telemetry_int` only had a `base` group table (since up to this point it only required one extra table), so with a new table we probably want to split this into `int_evpl` and `int_epl` since it'll still have to follow `mef_eline`'s table groups `evpl` and `epl` which currently default to table 0 anyway.

File: docs/blueprints/EP031.rst
Comment (viniarck) [reply docs/blueprints/EP031.rst:295]: Whatever the final decision on this related thread here https://github.com/kytos-ng/kytos/pull/402/files#r1276699359, it should probably follow the same behavior since ultimately `telemetry_int` is essentially "shadowing" `mef_eline` flows with higher priority ones and `mef_eline`'s is the one responsible for finding paths with the parametrized options.

File: docs/blueprints/EP031.rst
Comment (viniarck) [inline docs/blueprints/EP031.rst:323]: This requirement is still being discussed/refined here: https://github.com/kytos-ng/kytos/pull/402/files#r1276699359

File: docs/blueprints/EP031.rst
Comment (viniarck) [reply docs/blueprints/EP031.rst:295]: The main gotcha if it were to fall back and delete flows, is that consistency check wouldn't make sense from telemetry_int perspective to run again, since it'd be running the same consistency as mef_eline, and it can end up in a cycle in the worst case if a wrong action were to be installed. 

If a fallback to mef_eline is desirable when INT consistency check fails after a few attempts, then it should probably just disable INT and update status & status_reason metadata accordingly too. Consistency check from a INT perspective will only fail if flows haven't been installed yet or if indeed flows have wrong match or actions. So, as long as it doesn't pre-maturely use a short interval it should perform relatively stable. Mentioning these points just so you all are aware and we can make a decision of the behavior you need implemented.

File: docs/blueprints/EP031.rst
Comment (viniarck) [reply docs/blueprints/EP031.rst:251]: Final verdict: proxy port link down will remove INT flows, and when proxy port link goes up if the EVC is still up then it'll deploy again. I've updated text on https://github.com/kytos-ng/kytos/pull/402/commits/a699943ca8326582898d230d4145db2d0f1b7849

File: docs/blueprints/EP031.rst
Comment (viniarck) [reply docs/blueprints/EP031.rst]: - EVPLs will use table 2 by default
- EPLs will use table 3 by default

Updated the text on https://github.com/kytos-ng/kytos/pull/402/commits/2096fc5da5e675562feba9df0b584acd2e4818c5

File: docs/blueprints/EP031.rst
Comment (viniarck) [reply docs/blueprints/EP031.rst]: Final verdict: consistency check will disable INT (disable and remove the flows) if it fails after a few attempts in this first version. Updated the text on https://github.com/kytos-ng/kytos/pull/402/commits/c2330dce57198b2588b2b277402da8b9d7dde937

**Comment by viniarck** [comment]
@jab1982 and @italovalcy thanks for your inputs regarding the requirements meeting. I've pushed three new commits https://github.com/kytos-ng/kytos/pull/402/commits/a699943ca8326582898d230d4145db2d0f1b7849, https://github.com/kytos-ng/kytos/pull/402/commits/c2330dce57198b2588b2b277402da8b9d7dde937, https://github.com/kytos-ng/kytos/pull/402/commits/2096fc5da5e675562feba9df0b584acd2e4818c5 updating the text based on the decisions on yesterday's meeting. In case you want to re-review again this PR will stay open for a few days before getting merged.

**Pull request merged**
